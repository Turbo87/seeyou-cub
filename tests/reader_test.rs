use seeyou_cub::{CubReader, CubStyle};

#[test]
fn parse_france_fixture() {
    let mut reader = CubReader::from_path("tests/fixtures/france_2024.07.02.cub")
        .expect("Failed to open fixture file");
    let mut warnings = Vec::new();

    let header = reader
        .read_header(&mut warnings)
        .expect("Failed to parse header");

    // Validate warnings
    assert_eq!(warnings.len(), 0);

    // Snapshot the header
    insta::assert_debug_snapshot!("header", header);

    // Parse all items
    let items: Vec<_> = reader
        .read_items(&header, &mut warnings)
        .expect("Failed to create item iterator")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse items");

    // Validate item count
    assert_eq!(items.len(), 1368);

    // Snapshot first few items
    insta::assert_debug_snapshot!("items_sample", &items[0..5]);

    // Validate first item's points
    let first_item = items.first().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, first_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("first_item_points", points);

    let last_item = items.last().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, last_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("last_item_points", points);

    // Validate all airspaces and collect statistics
    let mut total_points = 0;
    let mut style_counts = std::collections::HashMap::new();

    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;

        let mut point_iter = reader
            .read_points(&header, item, &mut warnings)
            .expect("Failed to read points");

        for point_result in &mut point_iter {
            let _point = point_result.expect("Failed to parse point");
            total_points += 1;
        }
    }

    // Validate total points
    assert_eq!(total_points, 93784, "Total points mismatch");

    // Validate airspace type distribution
    assert_eq!(
        *style_counts.get(&CubStyle::Unknown).unwrap_or(&0),
        452,
        "Unknown airspace count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::RestrictedArea).unwrap_or(&0),
        435,
        "Restricted area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ProhibitedArea).unwrap_or(&0),
        113,
        "Prohibited area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::DangerArea).unwrap_or(&0),
        111,
        "Danger area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::GliderSector).unwrap_or(&0),
        135,
        "Glider sector count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ControlZone).unwrap_or(&0),
        92,
        "Control zone count mismatch"
    );
}
