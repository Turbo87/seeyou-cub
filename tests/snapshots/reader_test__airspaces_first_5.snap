---
source: tests/reader_test.rs
expression: "&airspaces[0..5]"
---
[
    Airspace {
        bounding_box: BoundingBox {
            left: 0.033180647,
            top: 0.83466476,
            right: 0.033224326,
            bottom: 0.8346504,
        },
        style: RestrictedArea,
        class: Unknown,
        extended_type: None,
        min_alt: 0,
        max_alt: 488,
        min_alt_style: AboveGroundLevel,
        max_alt_style: MeanSeaLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.83465517, lon: 0.033180647 },
            Point { lat: 0.8346504, lon: 0.033224326 },
            Point { lat: 0.83466476, lon: 0.03319982 },
            Point { lat: 0.83465517, lon: 0.033180647 },
        ],
        name: Some(
            "R265 LA GREMUSE",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        bounding_box: BoundingBox {
            left: 0.039682,
            top: 0.85236555,
            right: 0.039759766,
            bottom: 0.85232186,
        },
        style: ProhibitedArea,
        class: Unknown,
        extended_type: None,
        min_alt: 0,
        max_alt: 610,
        min_alt_style: AboveGroundLevel,
        max_alt_style: MeanSeaLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.85236555, lon: 0.039696384 },
            Point { lat: 0.8523362, lon: 0.039682 },
            Point { lat: 0.85232186, lon: 0.039744854 },
            Point { lat: 0.8523511, lon: 0.039759766 },
            Point { lat: 0.85236555, lon: 0.039696384 },
        ],
        name: Some(
            "P47 BALARD",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        bounding_box: BoundingBox {
            left: 0.099705316,
            top: 0.7667715,
            right: 0.099785745,
            bottom: 0.76671344,
        },
        style: ProhibitedArea,
        class: Unknown,
        extended_type: None,
        min_alt: 0,
        max_alt: 5944,
        min_alt_style: AboveGroundLevel,
        max_alt_style: FlightLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.7667715, lon: 0.09974686 },
            Point { lat: 0.7667715, lon: 0.099745795 },
            Point { lat: 0.7667715, lon: 0.0997442 },
            Point { lat: 0.7667715, lon: 0.0997426 },
            Point { lat: 0.7667715, lon: 0.09974153 },
            Point { lat: 0.7667715, lon: 0.09973994 },
            Point { lat: 0.76677096, lon: 0.09973834 },
            Point { lat: 0.76677096, lon: 0.09973727 },
            Point { lat: 0.76677096, lon: 0.09973568 },
            Point { lat: 0.7667704, lon: 0.09973461 },
            Point { lat: 0.7667704, lon: 0.09973302 },
            Point { lat: 0.7667699, lon: 0.09973195 },
            Point { lat: 0.76676935, lon: 0.09973035 },
            Point { lat: 0.76676935, lon: 0.099729285 },
            Point { lat: 0.7667688, lon: 0.09972769 },
            Point { lat: 0.7667683, lon: 0.099726625 },
            Point { lat: 0.7667678, lon: 0.09972556 },
            Point { lat: 0.76676726, lon: 0.099724494 },
            Point { lat: 0.7667667, lon: 0.09972289 },
            Point { lat: 0.7667662, lon: 0.09972183 },
            Point { lat: 0.76676565, lon: 0.09972076 },
            Point { lat: 0.7667646, lon: 0.099719696 },
            Point { lat: 0.76676404, lon: 0.09971863 },
            Point { lat: 0.7667635, lon: 0.099717565 },
            Point { lat: 0.766763, lon: 0.0997165 },
            Point { lat: 0.7667619, lon: 0.099715434 },
            Point { lat: 0.76676136, lon: 0.099714905 },
            Point { lat: 0.7667603, lon: 0.09971384 },
            Point { lat: 0.76675975, lon: 0.099712774 },
            Point { lat: 0.76675874, lon: 0.09971224 },
            Point { lat: 0.7667582, lon: 0.09971117 },
            Point { lat: 0.76675713, lon: 0.09971064 },
            Point { lat: 0.76675606, lon: 0.09971011 },
            Point { lat: 0.7667555, lon: 0.09970958 },
            Point { lat: 0.76675445, lon: 0.09970904 },
            Point { lat: 0.7667534, lon: 0.099707976 },
            Point { lat: 0.7667523, lon: 0.099707976 },
            Point { lat: 0.76675177, lon: 0.09970745 },
            Point { lat: 0.76675075, lon: 0.09970691 },
            Point { lat: 0.7667497, lon: 0.09970638 },
            Point { lat: 0.7667486, lon: 0.09970638 },
            Point { lat: 0.76674753, lon: 0.099705845 },
            Point { lat: 0.76674646, lon: 0.099705845 },
            Point { lat: 0.7667454, lon: 0.099705316 },
            Point { lat: 0.7667443, lon: 0.099705316 },
            Point { lat: 0.7667438, lon: 0.099705316 },
            Point { lat: 0.7667427, lon: 0.099705316 },
            Point { lat: 0.7667417, lon: 0.099705316 },
            Point { lat: 0.7667406, lon: 0.099705316 },
            Point { lat: 0.76673955, lon: 0.099705316 },
            Point { lat: 0.7667385, lon: 0.099705845 },
            Point { lat: 0.7667374, lon: 0.099705845 },
            Point { lat: 0.7667363, lon: 0.09970638 },
            Point { lat: 0.76673526, lon: 0.09970638 },
            Point { lat: 0.7667347, lon: 0.09970691 },
            Point { lat: 0.7667337, lon: 0.09970745 },
            Point { lat: 0.76673263, lon: 0.099707976 },
            Point { lat: 0.76673156, lon: 0.099707976 },
            Point { lat: 0.7667305, lon: 0.09970904 },
            Point { lat: 0.76672995, lon: 0.09970958 },
            Point { lat: 0.7667289, lon: 0.09971011 },
            Point { lat: 0.7667278, lon: 0.09971064 },
            Point { lat: 0.76672727, lon: 0.09971117 },
            Point { lat: 0.7667262, lon: 0.09971224 },
            Point { lat: 0.7667252, lon: 0.099712774 },
            Point { lat: 0.76672465, lon: 0.09971384 },
            Point { lat: 0.7667236, lon: 0.099714905 },
            Point { lat: 0.76672304, lon: 0.099715434 },
            Point { lat: 0.7667225, lon: 0.0997165 },
            Point { lat: 0.7667214, lon: 0.099717565 },
            Point { lat: 0.7667209, lon: 0.09971863 },
            Point { lat: 0.76672035, lon: 0.099719696 },
            Point { lat: 0.7667198, lon: 0.09972076 },
            Point { lat: 0.76671875, lon: 0.09972183 },
            Point { lat: 0.7667182, lon: 0.09972289 },
            Point { lat: 0.7667177, lon: 0.099724494 },
            Point { lat: 0.7667172, lon: 0.09972556 },
            Point { lat: 0.76671666, lon: 0.099726625 },
            Point { lat: 0.7667161, lon: 0.09972769 },
            Point { lat: 0.7667161, lon: 0.099729285 },
            Point { lat: 0.7667156, lon: 0.09973035 },
            Point { lat: 0.76671505, lon: 0.09973195 },
            Point { lat: 0.76671505, lon: 0.09973302 },
            Point { lat: 0.7667145, lon: 0.09973461 },
            Point { lat: 0.7667145, lon: 0.09973568 },
            Point { lat: 0.766714, lon: 0.09973727 },
            Point { lat: 0.766714, lon: 0.09973834 },
            Point { lat: 0.766714, lon: 0.09973994 },
            Point { lat: 0.76671344, lon: 0.09974153 },
            Point { lat: 0.76671344, lon: 0.0997426 },
            Point { lat: 0.76671344, lon: 0.0997442 },
            Point { lat: 0.76671344, lon: 0.099745795 },
            Point { lat: 0.76671344, lon: 0.09974686 },
            Point { lat: 0.76671344, lon: 0.09974846 },
            Point { lat: 0.76671344, lon: 0.09974953 },
            Point { lat: 0.766714, lon: 0.09975112 },
            Point { lat: 0.766714, lon: 0.099752724 },
            Point { lat: 0.766714, lon: 0.09975379 },
            Point { lat: 0.7667145, lon: 0.099755384 },
            Point { lat: 0.7667145, lon: 0.09975645 },
            Point { lat: 0.76671505, lon: 0.09975805 },
            Point { lat: 0.76671505, lon: 0.09975912 },
            Point { lat: 0.7667156, lon: 0.09976071 },
            Point { lat: 0.7667161, lon: 0.09976178 },
            Point { lat: 0.7667161, lon: 0.09976338 },
            Point { lat: 0.76671666, lon: 0.09976444 },
            Point { lat: 0.7667172, lon: 0.0997655 },
            Point { lat: 0.7667177, lon: 0.099767104 },
            Point { lat: 0.7667182, lon: 0.09976817 },
            Point { lat: 0.76671875, lon: 0.099769235 },
            Point { lat: 0.7667198, lon: 0.0997703 },
            Point { lat: 0.76672035, lon: 0.099771366 },
            Point { lat: 0.7667209, lon: 0.09977243 },
            Point { lat: 0.7667214, lon: 0.0997735 },
            Point { lat: 0.7667225, lon: 0.09977456 },
            Point { lat: 0.76672304, lon: 0.09977563 },
            Point { lat: 0.7667236, lon: 0.099776156 },
            Point { lat: 0.76672465, lon: 0.09977722 },
            Point { lat: 0.7667252, lon: 0.09977829 },
            Point { lat: 0.7667262, lon: 0.09977882 },
            Point { lat: 0.76672727, lon: 0.09977989 },
            Point { lat: 0.7667278, lon: 0.09978042 },
            Point { lat: 0.7667289, lon: 0.099780954 },
            Point { lat: 0.76672995, lon: 0.09978148 },
            Point { lat: 0.7667305, lon: 0.09978255 },
            Point { lat: 0.76673156, lon: 0.099783085 },
            Point { lat: 0.76673263, lon: 0.099783614 },
            Point { lat: 0.7667337, lon: 0.099783614 },
            Point { lat: 0.7667347, lon: 0.09978415 },
            Point { lat: 0.76673526, lon: 0.09978468 },
            Point { lat: 0.7667363, lon: 0.09978468 },
            Point { lat: 0.7667374, lon: 0.099785216 },
            Point { lat: 0.7667385, lon: 0.099785216 },
            Point { lat: 0.76673955, lon: 0.099785745 },
            Point { lat: 0.7667406, lon: 0.099785745 },
            Point { lat: 0.7667417, lon: 0.099785745 },
            Point { lat: 0.7667427, lon: 0.099785745 },
            Point { lat: 0.7667438, lon: 0.099785745 },
            Point { lat: 0.7667443, lon: 0.099785745 },
            Point { lat: 0.7667454, lon: 0.099785745 },
            Point { lat: 0.76674646, lon: 0.099785216 },
            Point { lat: 0.76674753, lon: 0.099785216 },
            Point { lat: 0.7667486, lon: 0.09978468 },
            Point { lat: 0.7667497, lon: 0.09978468 },
            Point { lat: 0.76675075, lon: 0.09978415 },
            Point { lat: 0.76675177, lon: 0.099783614 },
            Point { lat: 0.7667523, lon: 0.099783614 },
            Point { lat: 0.7667534, lon: 0.099783085 },
            Point { lat: 0.76675445, lon: 0.09978255 },
            Point { lat: 0.7667555, lon: 0.09978148 },
            Point { lat: 0.76675606, lon: 0.099780954 },
            Point { lat: 0.76675713, lon: 0.09978042 },
            Point { lat: 0.7667582, lon: 0.09977989 },
            Point { lat: 0.76675874, lon: 0.09977882 },
            Point { lat: 0.76675975, lon: 0.09977829 },
            Point { lat: 0.7667603, lon: 0.09977722 },
            Point { lat: 0.76676136, lon: 0.099776156 },
            Point { lat: 0.7667619, lon: 0.09977563 },
            Point { lat: 0.766763, lon: 0.09977456 },
            Point { lat: 0.7667635, lon: 0.0997735 },
            Point { lat: 0.76676404, lon: 0.09977243 },
            Point { lat: 0.7667646, lon: 0.099771366 },
            Point { lat: 0.76676565, lon: 0.0997703 },
            Point { lat: 0.7667662, lon: 0.099769235 },
            Point { lat: 0.7667667, lon: 0.09976817 },
            Point { lat: 0.76676726, lon: 0.099767104 },
            Point { lat: 0.7667678, lon: 0.0997655 },
            Point { lat: 0.7667683, lon: 0.09976444 },
            Point { lat: 0.7667688, lon: 0.09976338 },
            Point { lat: 0.76676935, lon: 0.09976178 },
            Point { lat: 0.76676935, lon: 0.09976071 },
            Point { lat: 0.7667699, lon: 0.09975912 },
            Point { lat: 0.7667704, lon: 0.09975805 },
            Point { lat: 0.7667704, lon: 0.09975645 },
            Point { lat: 0.76677096, lon: 0.099755384 },
            Point { lat: 0.76677096, lon: 0.09975379 },
            Point { lat: 0.76677096, lon: 0.099752724 },
            Point { lat: 0.7667715, lon: 0.09975112 },
            Point { lat: 0.7667715, lon: 0.09974953 },
            Point { lat: 0.7667715, lon: 0.09974846 },
            Point { lat: 0.7667715, lon: 0.09974686 },
            Point { lat: 0.7667715, lon: 0.099745795 },
        ],
        name: Some(
            "D595 Laser",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        bounding_box: BoundingBox {
            left: 0.093006656,
            top: 0.8027499,
            right: 0.09311798,
            bottom: 0.8026769,
        },
        style: RestrictedArea,
        class: Unknown,
        extended_type: None,
        min_alt: 0,
        max_alt: 1067,
        min_alt_style: AboveGroundLevel,
        max_alt_style: MeanSeaLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.8027499, lon: 0.09304554 },
            Point { lat: 0.8026769, lon: 0.093006656 },
            Point { lat: 0.8026769, lon: 0.09311798 },
            Point { lat: 0.8027499, lon: 0.09309401 },
            Point { lat: 0.8027499, lon: 0.09304554 },
        ],
        name: Some(
            "LF R 258 AMBERIEU",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
    Airspace {
        bounding_box: BoundingBox {
            left: 0.03643375,
            top: 0.851842,
            right: 0.03662763,
            bottom: 0.8517546,
        },
        style: RestrictedArea,
        class: Unknown,
        extended_type: None,
        min_alt: 0,
        max_alt: 457,
        min_alt_style: AboveGroundLevel,
        max_alt_style: MeanSeaLevel,
        time_out: 0,
        start_date: None,
        end_date: None,
        extra_data: 0,
        days_active: DaysActive {
            bits: 0,
        },
        points: [
            Point { lat: 0.851842, lon: 0.03662763 },
            Point { lat: 0.85182226, lon: 0.03648222 },
            Point { lat: 0.8517546, lon: 0.03643375 },
            Point { lat: 0.8517642, lon: 0.036579162 },
            Point { lat: 0.851842, lon: 0.03662763 },
        ],
        name: Some(
            "LF R 84 B DOMAINE DE VERSAILLES",
        ),
        frequency_name: None,
        icao_code: None,
        exception_rules: None,
        notam_remarks: None,
        notam_id: None,
        frequency: None,
        secondary_frequency: None,
        notam_insert_time: None,
    },
]
